### ISSUE TITLE
After 'swarm init', unable to create network using local-scoped ipam plugin : "weavemesh"

### ISSUE THREAD

**Issue by nkratzke** [docker/docker#23990]
I just tried to launch Weave with Docker Swarm (Version 1.12, RC 2).
That is not working.

According to Weave engineers this is due to that Docker is not supporting plugins in Swarm mode, (Version 1.12). https://github.com/docker/libnetwork/issues/1296#issuecomment-228572727

However, this leads to the following situation, that weave (and other network plugins) can not be used with Docker swarm after Docker 1.12.

I do not think, that this is a wishful situation for the community. Plugins are really useless then ...

**Comment by cpuguy83** [comment]
@nkratzke network plugins still work as they always have, just not with services (yet).
I would have to defer to the libnetwork folks as to the challenges of integrating network plugins support with services.

**Comment by rade** [comment]
> network plugins still work as they always have

Not once you have issued a `docker swarm init`. See weaveworks/weave#2403.

**Comment by cpuguy83** [comment]
I'm not sure what should be expected here...
I'm not surprised that since plugins are not supported in swarm mode, that once the node joins a swarm (via `docker swarm join` or `docker swarm init`) that it can't use plugins... but will again defer to the libnetwork folks on if that's expected or not.

Note that `swarm mode` is not the same as using github.com/docker/swarm

ping @mavenugo

**Comment by cpuguy83** [comment]
I also noted on the referenced issue that there seems to be some in-memory state that doesn't get cleared after leaving the swarm. I feel like that's a bug.

**Comment by rade** [comment]
> I also noted on the referenced issue that there seems to be some in-memory state that doesn't get cleared after leaving the swarm. I feel like that's a bug.

Quite. The other strange symptom documented in that issue is that the complaint about the "invalid ipam driver: weavemesh" goes away when said driver has been used prior to `docker swarm init`. So it looks like once in swarm mode, docker no longer finds *new* plugin drivers but can continue to talk to pre-existing ones.

**Comment by mavenugo** [comment]
@rade @nkratzke @cpuguy83 network plugins are not currently supported in swarmkit. When swarm-mode is enabled, the global network creation is handled by swarmkit and hence the lack of plugin support affects the workflow. I think we should error out cleanly in swarmkit when the network is created.
cc @mrjana

**Comment by rade** [comment]
> the global network creation is handled by swarmkit

What about local-mode plugins? We are seeing these errors there too.

**Comment by mrjana** [comment]
@mavenugo I think this is about local scope plugins. To clarify all local scope plugins should continue to work even if swarm mode is enabled. So there seems to be a bug in the engine-api implementation in docker where a local scope plugin is affected when swarm mode is enabled. We should fix this before 1.12 release

**Comment by rade** [comment]
> I think this is about local scope plugins.

The `weave launch` in the examples in weaveworks/weave#2403 does indeed (attempt to) create a network from a local scope driver.

**Comment by mavenugo** [comment]
@rade could you pls point us to the code that implements the `weave launch` wrapper ? It isn't clear on why the failure is seen on the ipam-driver (as mentioned in https://github.com/weaveworks/weave/issues/2403)

**Comment by rade** [comment]
> code that implements the `weave launch` wrapper

https://github.com/weaveworks/weave/blob/v1.6.0/weave#L1940

That is going to be quite hard to follow. The bit that provokes the error is in https://github.com/weaveworks/weave/blob/v1.6.0/weave#L1768, which is implemented in https://github.com/weaveworks/weave/blob/v1.6.0/prog/weaveutil/plugin_network.go#L12 and essentially boils down to the equivalent of `docker network create --driver=weavemesh --ipam-driver=weavemesh weave`.

**Comment by bboreham** [comment]
Some logs, in case it helps, from running the command:

    docker network create --driver=weavemesh --ipam-driver=weavemesh foo

Docker logs in debug mode:

````
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.101653132Z" level=debug msg="Calling POST /v1.24/networks/create"
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.102107151Z" level=debug msg="form data: {\"CheckDuplicate\":true,\"Driver\":\"weavemesh\",\"EnableIPv6\":false,\"IPAM\":{\"Config\":[],\"Driver\":\"weavemesh\",\"Options\":{}},\"Internal\":false,\"Labels\":{},\"Name\":\"foo\",\"Options\":{}}"
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.104659577Z" level=info msg="remote ipam driver weavemesh does not support capabilities"
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.105141762Z" level=debug msg="IpamDriver.GetCapabilities: 404 page not found\n"
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.105856052Z" level=debug msg="weavemesh implements: NetworkDriver"
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.106218969Z" level=debug msg="weavemesh implements: IpamDriver"
Jun 27 18:50:00 host1 docker[7907]: time="2016-06-27T18:50:00.106578243Z" level=error msg="Handler for POST /v1.24/networks/create returned error: invalid ipam driver: \"weavemesh\""
````

From our own plugin logs I see `GetDefaultAddressSpaces` being called, so that leads me to believe that `DrvRegistry.registerIpamDriver` is erroring after that.

**Comment by hkvemuri** [comment]
I guess the same is true for volume plugins ?

**Comment by mavenugo** [comment]
@hkvemuri this issue is not related to volume plugins.

**Comment by mavenugo** [comment]
@nkratzke @bboreham we should narrow down the issue to the `ipam-driver` local-scope plugin error. Swarm-mode not supporting global driver is a known issue and must be addressed independently.

To that effect, I will change the issue title to reflect the reality as we discussed in https://github.com/weaveworks/weave/issues/2403

**Comment by mavenugo** [comment]
@bboreham @nkratzke  FYI, we decided to not go with #24172 which you originally validated.
But with #23312 merged, it should address your problem. Can you please confirm ?

**Comment by bboreham** [comment]
For my understanding, is it https://github.com/docker/libnetwork/pull/1310 that is the actual fix?

**Comment by nkratzke** [comment]
@mavenugo I think @bboreham is the expert here. I just stumbled over the issue and triggered the initial ticket. I am no expert in the details. So, @bboreham has the more relevant opinion here.

**Comment by mavenugo** [comment]
@bboreham https://github.com/docker/libnetwork/pull/1310 and https://github.com/docker/swarmkit/pull/1130 are the fixes.

**Comment by bboreham** [comment]
OK, @mavenugo, I pulled docker at master, built it, ran `docker swarm init` then ran the `weave launch` command, and no error messages.  Looks OK, thanks!

**Comment by mavenugo** [comment]
Thanks for confirming. Will close this Issue.

**Comment by thaJeztah** [comment]
@bboreham thanks, that's good to hear!
